File: app/src/main/java/com/kite/mnemoai/data/repository/GroupRepository.java
package com.kite.mnemoai.data.repository;

import androidx.lifecycle.LiveData;

import com.kite.mnemoai.data.local.dao.GroupDao;
import com.kite.mnemoai.data.local.dao.WordGroupDao;
import com.kite.mnemoai.data.local.entity.GroupEntity;
import com.kite.mnemoai.data.local.entity.WordGroupEntity;
import com.kite.mnemoai.data.model.GroupDetail;
import com.kite.mnemoai.ui.vocabularybook.AllVocabularyBookItem;
import com.kite.mnemoai.ui.vocabularybook.LearningVocabularyBookItem;

import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.stream.Collectors;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class GroupRepository {
    private GroupDao groupDao;
    private WordGroupDao wordGroupDao;

    private ExecutorService executors;

    @Inject
    public GroupRepository(ExecutorService executors, GroupDao groupDao, WordGroupDao wordGroupDao) {
        this.executors = executors;
        this.groupDao = groupDao;
        this.wordGroupDao = wordGroupDao;
    }

    public LiveData<List<GroupDetail>> getAllGroupLiveData(){
        return groupDao.getAllGroupDeatilsLiveData();
    };

    public LiveData<GroupEntity> getGroupLiveDataById(long id){
        return groupDao.getGroupEntityById(id);
    }

    public LiveData<List<GroupDetail>> getGroupDetailLiveDataById(long id){
        return groupDao.getGroupDetailLiveDataById(id);
    }

    public void getAllUnlearningGroups(IRepositoryCallback<List<GroupEntity>> callback){
        executors.execute(() -> {
            List<GroupEntity> groupEntities = groupDao.getAllUnlearningGroupEntities();
            callback.onComplete(groupEntities);
        });
    }

    public void addLearningGroups(List<Long> ids){
        executors.execute(() -> {
            groupDao.addLearningGroups(ids);
        });
    }

    public void setVocabularyBookLearningStatus(Long id, int status){
        executors.execute(() -> {
            groupDao.setVocabularyBookLearningStatus(id, status);
        });
    }

    public void addNewOrModifyVocabularyBook(GroupEntity groupEntity){
        executors.execute(() -> groupDao.insertGroup(groupEntity));
    }

    public void modifyVocabularyBook(GroupEntity groupEntity){
        executors.execute(() -> groupDao.updateGroup(groupEntity));
    }

    public void deleteGroup(long id){
        executors.execute(() -> groupDao.deleteGroupById(id));
    }

    public void addAlterWords(long groupId, List<Long> ids){
        executors.execute(() -> {
            List<WordGroupEntity> wordGroupEntities = ids.stream().map(id -> new WordGroupEntity(id, groupId)).collect(Collectors.toList());
            wordGroupDao.insertWordGroups(wordGroupEntities);
        });
    }

    public void removeAlterWords(long groupId, List<Long> ids){
        executors.execute(() -> wordGroupDao.deleteWordGroupItem(groupId, ids));
    }

    public LiveData<List<LearningVocabularyBookItem>> getLearningVocabularyBookItem(){
        return groupDao.getLearningVocabularyBookItem();
    }

    public LiveData<List<AllVocabularyBookItem>> getAllVocabularyBookItem(){
        return groupDao.getAllVocabularyBookItem();
    }
}

File: core/model/src/main/java/com/kite/mnemoai/model/repository/GroupRepository.kt
package com.kite.mnemoai.model.repository

import com.kite.mnemoai.model.Result
import com.kite.mnemoai.model.group.AllVocabularyBookItem
import com.kite.mnemoai.model.group.Group
import com.kite.mnemoai.model.group.GroupDetail
import com.kite.mnemoai.model.group.LearningVocabularyBookItem
import kotlinx.coroutines.flow.Flow

interface GroupRepository {
    fun observeAllGroups(): Flow<Result<List<GroupDetail>>>

    fun observeGroupById(id: Long): Flow<Result<Group?>>

    fun observeGroupDetailById(id: Long): Flow<Result<List<GroupDetail>>>

    suspend fun getAllUnlearningGroups(): Result<List<Group>>

    suspend fun addLearningGroups(ids: List<Long>)

    suspend fun setVocabularyBookLearningStatus(id: Long, status: Int)

    suspend fun addNewOrModifyVocabularyBook(group: Group)

    suspend fun modifyVocabularyBook(group: Group)

    suspend fun deleteGroup(id: Long)

    suspend fun addAlterWords(groupId: Long, ids: List<Long>)

    suspend fun removeAlterWords(groupId: Long, ids: List<Long>)

    fun observeLearningVocabularyBookItems(): Flow<Result<List<LearningVocabularyBookItem>>>

    fun observeAllVocabularyBookItems(): Flow<Result<List<AllVocabularyBookItem>>>
}
